File: find-ami.go
package main


import (
	"context"
	"fmt"
	"errors"
	"sort"

	"github.com/aws/aws-lambda-go/cfn"
	"github.com/aws/aws-lambda-go/lambda"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/ec2"
	"github.com/google/uuid"
)


// Extract a string from the given `event`. Returns the empty string
// in case of any error.
func getString(event *cfn.Event, key string) (value string) {
	value = ""
	defer func() {
		recover()
	}()
	value = event.ResourceProperties[key].(string)
	return
}


// Lambda event handler
func findAmi(ctx context.Context, event cfn.Event) (physicalId string, data map[string]interface{}, err error) {
	if event.RequestType == "Delete" {
		fmt.Println("Request type is 'Delete'; nothing to do")
		return
	}
	fmt.Printf("Request type is '%v'\n", event.RequestType)

	// Generate a random physial ID to keep CloudFormation happy
	physicalId = uuid.New().String()  // Random UUID

	// Extract filter values from the `ResourceProperties`
	region := getString(&event, "Region")
	if region == "" {
		err = errors.New("'Region' must be set in the request's ResourceProperties")
		fmt.Println(err)
		return
	}
	debug := getString(&event, "Debug")
	architecture := getString(&event, "Architecture")
	name := getString(&event, "Name")
	ownerId := getString(&event, "OwnerId")
	rootDeviceType := getString(&event, "RootDeviceType")
	if rootDeviceType == "" {
		rootDeviceType = "ebs"
	}
	virtualizationType := getString(&event, "VirtualizationType")
	if virtualizationType == "" {
		virtualizationType = "hvm"
	}
	if debug == "true" {
		fmt.Println("Requested debug: true")
		fmt.Println("Requested region:", region)
		fmt.Println("Requested architecture:", architecture)
		fmt.Println("Requested name:", name)
		fmt.Println("Requested owner id:", ownerId)
		fmt.Println("Requested root device type:", rootDeviceType)
	}

	// Query the EC2 service to list the matching AMIs
	cfg := &aws.Config{
		Region: &region,
	}
	sess := session.Must(session.NewSession(cfg))
	ec2Service := ec2.New(sess)
	input := new(ec2.DescribeImagesInput)
	if architecture != "" {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name: aws.String("architecture"),
			Values: []*string{&architecture},
		})
	}
	if name != "" {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name: aws.String("name"),
			Values: []*string{&name},
		})
	}
	if ownerId != "" {
		input.Filters = append(input.Filters, &ec2.Filter{
			Name: aws.String("owner-id"),
			Values: []*string{&ownerId},
		})
	}
	input.Filters = append(input.Filters, &ec2.Filter{
		Name: aws.String("root-device-type"),
		Values: []*string{&rootDeviceType},
	})
	input.Filters = append(input.Filters, &ec2.Filter{
		Name: aws.String("virtualization-type"),
		Values: []*string{&virtualizationType},
	})
	output, err := ec2Service.DescribeImages(input)
	if err != nil {
		fmt.Printf("API call to EC2.DescribeImages() failed: %v\n", err)
		return
	}
	fmt.Println("API call to EC2.DescribeImages() succeeded")
	fmt.Printf("%d images found", len(output.Images))
	if len(output.Images) == 0 {
		errorString := "No image found for the given filters"
		err = errors.New(errorString)
		return
	}

	// Sort the matching AMIs by creation date and return the latest one
	sort.Slice(output.Images, func(i, j int) bool {
		return *(output.Images[i].CreationDate) > *(output.Images[j].CreationDate)
	})
	data = map[string]interface{} {
		"Id": *(output.Images[0].ImageId),
		"Name": *(output.Images[0].Name),
		"Description": *(output.Images[0].Description),
	}
	return
}


func main() {
	lambda.Start(cfn.LambdaWrap(findAmi))
}

File: find-ami.py
#!/usr/bin/env python3

import boto3
import uuid
import json
import requests


def handler(event, context):
    """
    Handle an operation for a "FindAmi" CloudFormation custom resource.

    The received event has the following pattern:

    {
       "RequestType" : "Create",
       "ResponseURL" : "http://pre-signed-S3-url-for-response",
       "StackId" : "arn:aws:cloudformation:us-west-2:123456789012:stack/stack-name/guid",
       "RequestId" : "unique id for this create request",
       "ResourceType" : "Custom::TestResource",
       "LogicalResourceId" : "MyTestResource",
       "ResourceProperties" : {
          "Name" : "Value",
          "List" : [ "1", "2", "3" ]
       }
    }
    """
    if event['RequestType'] == "Delete":
        print(f"Request type is 'Delete'; nothing to do")
        send_response(event, True)
        return
    print(f"Request type is '{event['RequestType']}'")

    prop = event['ResourceProperties']
    region = prop.get('Region')
    if not region:
        msg = f"'Region' must be set in the request's ResourceProperties"
        print(msg)
        send_response(event, False, msg)
        return

    debug = prop.get('Debug')
    if debug == "true":
        debug = True
    else:
        debug = False

    architecture = prop.get('Architecture', "")
    name = prop.get('Name', "")
    owner_id = prop.get('OwnerId', "")
    root_device_type = prop.get('RootDeviceType', "ebs")
    virtualization_type = prop.get('VirtualizationType', "hvm")

    if debug:
        print(f"Requested debug: true")
        print(f"Requested region: {region}")
        print(f"Requested architecture: {architecture}")
        print(f"Requested name: {name}")
        print(f"Requested owner id: {owner_id}")
        print(f"Requested root device type: {root_device_type}")
        print(f"Requested virtualization type: {virtualization_type}")

    ec2 = boto3.client("ec2", region_name=region)

    filters = []
    if architecture:
        filters.append({'Name': "architecture", 'Values': [architecture]})
    if name:
        filters.append({'Name': "name", 'Values': [name]})
    if owner_id:
        filters.append({'Name': "owner-id", 'Values': [owner_id]})
    filters.append({'Name': "root-device-type", 'Values': [root_device_type]})
    filters.append({'Name': "virtualization-type", 'Values': [virtualization_type]})

    response = ec2.describe_images(Filters=filters)
    images = response['Images']
    if not images:
        msg = f"No image found for filters: {filters}"
        print(f"ERROR: {msg}")
        send_response(event, False, msg)
        return
    print(f"API call to ec2.describe_images() succeeded; {len(images)} images found")

    # Sort the matching AMIs by creation date and return the latest one
    sorted_images = sorted(images, key=lambda i: i['CreationDate'], reverse=True)
    image = sorted_images[0]
    data = {
        'Id': image['ImageId'],
        'Name': image['Name'],
        'Description': image['Description']
    }
    print(f"Latest image: {data}")

    send_response(event, True, "Success", data)


def send_response(event, success: bool, msg="", data={}):
    response = {
        'Status': "SUCCESS" if success else "FAILED",
        'Reason': msg,
        'PhysicalResourceId': str(uuid.uuid4()),
        'StackId': event['StackId'],
        'RequestId': event['RequestId'],
        'LogicalResourceId': event['LogicalResourceId'],
        'Data': data
    }
    data = json.dumps(response)
    requests.put(event['ResponseURL'], data=data)
